Add KIC Gateway API Guide (#3622)
Kong/docs.konghq.com: app/kubernetes-ingress-controller/2.2.x/guides/using-gateway-api.md

support backends other than Service, and manage the proxies that implement
routes.

Gateway API and Kong's implementation of Gateway API are both alpha stage and
under active development. Features and implementation specifics will change
before their initial general availability release.

## Enabling the feature

The Gateway API CRDs are not yet available by default in Kubernetes. You must
first [install them](https://gateway-api.sigs.k8s.io/v1alpha2/guides/getting-started/#installing-gateway-api-crds-manually).

The default controller configuration disables Gateway API handling. To enable
it, set `ingressController.env.feature_gates: Gateway=true` in your Helm
values.yaml (set `CONTROLLER_FEATURE_GATES=Gateway=true` if not using Helm).
Note that you must restart Pods with this flag set _after_ installing the
Gateway API CRDs.

If using Helm, you must use chart version 2.7 or higher. Older versions do not
include the ServiceAccount permissions necessary for KIC to read Gateway API
resources.

## Testing Connectivity to Kong

This guide assumes that the `PROXY_IP` environment variable is
set to contain the IP address or URL pointing to Kong.

> Please follow one of the

Reviewer: ```suggestion
Follow one of the
```

[deployment guides](/kubernetes-ingress-controller/{{page.kong_version}}/deployment/overview) to configure this environment variable.

If everything is setup correctly, making a request to Kong should return
HTTP 404 Not Found.

{:.note}
> **Note**: If you are running the example using Minikube on MacOS, you may need 
to run [`minikube tunnel`](https://minikube.sigs.k8s.io/docs/handbook/accessing/#loadbalancer-access)
in a separate terminal window.  This exposes LoadBalancer services 
externally, which is not enabled by default.

```bash
$ curl -i $PROXY_IP
HTTP/1.1 404 Not Found
Content-Type: application/json; charset=utf-8
Connection: keep-alive
Content-Length: 48
Server: kong/1.2.1

{"message":"no Route matched with those values"}
```

This is expected as Kong does not yet know how to proxy the request.

## Set up an echo-server

> Setup an echo-server application to demonstrate how

Reviewer: ```suggestion
Set up an echo-server application to demonstrate how
```

to use the {{site.kic_product_name}}:

```bash
$ kubectl apply -f https://bit.ly/echo-service
```

## Add a GatewayClass and Gateway

The Gateway resource represents the proxy instance that handles traffic for a
set of Gateway API routes, and a GatewayClass describes characteristics shared
by all Gateways of a given type.

> 

Reviewer: ```suggestion
Add a GatewayClass:

```

```bash
$ echo "apiVersion: gateway.networking.k8s.io/v1alpha2
apiVersion: gateway.networking.k8s.io/v1alpha2
kind: GatewayClass
metadata:
  name: kong
spec:
  controllerName: konghq.com/kic-gateway-controller
" | kubectl apply -f -

> gatewayclass.gateway.networking.k8s.io/kong created

Reviewer: Please either comment out the response or put it in a separate codeblock. We keep getting feedback/complaints that users can't just copy/paste the command as-is due to the extra line of text.

Usually something like this works:

> Response:
> ```
> gatewayclass.gateway.networking.k8s.io/kong created
> ```

Applies to all examples in this guide.

```

> 

Reviewer: ```suggestion
Add a Gateway: 

```

resource, we set the `konghq.com/gateway-unmanaged` annotation to the
`<namespace>/<name>` of the Kong proxy Service. This instructs KIC to populate
that Gateway resource with listener and status information. You can check to
confirm if KIC has updated the bound Gateway by inspecting the list of
associated addresses:

```bash
$ kubectl get gateway kong -o=jsonpath='{.status.addresses}' | jq
[
  {
    "type": "IPAddress",
    "value": "10.96.179.122"
  },
  {
    "type": "IPAddress",
    "value": "10.96.179.122"
  },
  {
    "type": "IPAddress",
    "value": "172.18.0.240"
  }
]
```

## Add an HTTPRoute

HTTPRoute resources are similar to Ingress resources: they contain a set of
matching criteria for HTTP requests and upstream Services to route those
requests to.

```bash
$ echo "apiVersion: gateway.networking.k8s.io/v1alpha2
kind: HTTPRoute
metadata:
  name: echo
spec:
  parentRefs:
  - group: gateway.networking.k8s.io
    kind: Gateway
    name: kong
  rules:
  - backendRefs:
    - group: ""
      kind: Service
      name: echo
      port: 80
      weight: 1
    matches:
    - path:
        type: PathPrefix
        value: /echo
" | kubectl apply -f -
```

> After creating an HTTPRoute, accessing `/echo` will forward a request to the

Reviewer: ```suggestion
After creating an HTTPRoute, accessing `/echo` forwards a request to the
```
